### PR TITLE
[SPARK-15187] [SQL] Disallow Dropping Default Database

### PR DESCRIPTION
**Pull request by gatorsmile**
#### What changes were proposed in this pull request?

In Hive Metastore, dropping default database is not allowed. However, in `InMemoryCatalog`, this is allowed. 

This PR is to disallow users to drop default database. 
#### How was this patch tested?

Previously, we already have a test case in HiveDDLSuite. Now, we also add the same one in DDLSuite

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @yhuai @andrewor14 @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 1]
```diff
@@ -118,6 +118,9 @@ class SessionCatalog(
   }
 
   def dropDatabase(db: String, ignoreIfNotExists: Boolean, cascade: Boolean): Unit = {
+    if (db == "default") {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: how about case sensitivety?

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: You are asking a very great question! Actually, I am not very sure how it works. 

First, the current code base is not considering case sensitivity in database names. That part is missing. I think we should call `toLowerCase` for all the cases.

Second, I saw we have a function `formatTableName` for formatting table names based on the configuration of `CASE_SENSITIVE`. However, the underlying Hive metastore is not case sensitive. See the document: https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DDL

My question is how to achieve the case sensitivity when using Hive metastore? also cc @yhuai

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: Added test cases to show the differences between Hive metastore and `InMemoryCatalog`

- InMemoryCatalog: https://github.com/gatorsmile/spark/blob/dropDefaultDB/sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala#L943

- Hive Metastore:
https://github.com/gatorsmile/spark/blob/dropDefaultDB/sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala#L500

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: Now, I think we should issue an exception if users set `CASE_SENSITIVE` to `true` if users use Hive Metastore.

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: So far, the best way I found is to override `caseSensitiveAnalysis` to `false` in `conf` of `HiveSessionState`. 
```scala
  override lazy val conf: SQLConf = new SQLConf {
    override def caseSensitiveAnalysis: Boolean = false
  }
```

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: Let me submit a PR to fix it. : )

**Comment by gatorsmile** [comment]
@cloud-fan https://github.com/apache/spark/pull/12993 resolves the issue you mentioned above. Will change this PR too for resolving the issues of database names by calling `formatDatabaseName`

Thanks!

**Comment by andrewor14** [comment]
retest this please

**Comment by andrewor14** [comment]
LGTM

**Comment by gatorsmile** [comment]
To fix the test failure, we need to call `formatDatabaseName` on the database name in all the `SessionCatalog` and `HiveSessionCatalog` APIs. Will do it this afternoon. Thanks!

**Comment by gatorsmile** [comment]
@andrewor14 @cloud-fan The PR is ready for review. Thanks!

**Comment by cloud-fan** [comment]
LGTM , merging to master and 2.0, thanks!

**Pull request closed without merging**